File: crates/mithril-oxide-sys/cpp/Dialect/PDL/IR/PDLTypes.cpp
#include "PDLTypes.hpp"

#include <mlir/Dialect/PDL/IR/PDLTypes.h>
#include <mlir/IR/MLIRContext.h>
#include <mlir/IR/Types.h>


namespace mithril_oxide_sys::pdl {

const void *AttributeType_get(MLIRContext *ctx)
{
    return AttributeType::get(ctx).getAsOpaquePointer();
}

const void *OperationType_get(MLIRContext *ctx)
{
    return OperationType::get(ctx).getAsOpaquePointer();
}

const void *RangeType_get(const void *elementType)
{
    return RangeType::get(mlir::Type::getFromOpaquePointer(elementType)).getAsOpaquePointer();
}

const void *TypeType_get(MLIRContext *ctx)
{
    return TypeType::get(ctx).getAsOpaquePointer();
}

const void *ValueType_get(MLIRContext *ctx)
{
    return ValueType::get(ctx).getAsOpaquePointer();
}

} // namespace mithril_oxide_sys::pdl

File: crates/mithril-oxide-sys/src/Dialect/PDL/IR/PDLTypes.rs
pub use self::ffi::{
    AttributeType_get, OperationType_get, RangeType_get, TypeType_get, ValueType_get,
};

#[cxx::bridge]
mod ffi {
    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/PDL/IR/PDLTypes.hpp");

        type MLIRContext = crate::IR::MLIRContext::MLIRContext;
    }

    #[namespace = "mithril_oxide_sys"]
    unsafe extern "C++" {
        type c_void = crate::c_void;
    }

    #[namespace = "mithril_oxide_sys::pdl"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/PDL/IR/PDLTypes.hpp");

        pub unsafe fn AttributeType_get(context: *mut MLIRContext) -> *const c_void;
        pub unsafe fn OperationType_get(context: *mut MLIRContext) -> *const c_void;
        #[must_use]
        pub unsafe fn RangeType_get(elementType: *const c_void) -> *const c_void;
        pub unsafe fn TypeType_get(context: *mut MLIRContext) -> *const c_void;
        pub unsafe fn ValueType_get(context: *mut MLIRContext) -> *const c_void;
    }
}
